src/DI/ServiceLookup/ILEmitResolverBuilder.cs

```diff
@@ -0,0 +1,370 @@
+        private static readonly FieldInfo FactoriesField = typeof(ILEmitResolverBuilderRuntimeContext).GetField(nameof(ILEmitResolverBuilderRuntimeContext.Factories));
+        private static readonly FieldInfo ConstantsField = typeof(ILEmitResolverBuilderRuntimeContext).GetField(nameof(ILEmitResolverBuilderRuntimeContext.Constants));
+
+        private class ILEmitResolverBuilderRuntimeContext
+        {
+            public CallSiteRuntimeResolver RuntimeResolver;
+            public IServiceScopeFactory ScopeFactory;
+            public ServiceProviderEngineScope Root;
+            public object[] Constants;
+            public Func<IServiceProvider, object>[] Factories;
+        }
+
+        private readonly CallSiteRuntimeResolver _runtimeResolver;
+
+        private readonly IServiceScopeFactory _serviceScopeFactory;
+
+        private readonly ServiceProviderEngineScope _rootScope;
+
+        public ILEmitResolverBuilder(CallSiteRuntimeResolver runtimeResolver, IServiceScopeFactory serviceScopeFactory, ServiceProviderEngineScope rootScope)
+        {
+            if (runtimeResolver == null)
+            {
+                throw new ArgumentNullException(nameof(runtimeResolver));
+            }
+            _runtimeResolver = runtimeResolver;
+            _serviceScopeFactory = serviceScopeFactory;
+            _rootScope = rootScope;
+        }
+
+        public Func<ServiceProviderEngineScope, object> Build(IServiceCallSite callSite)
+        {
+            if (callSite is SingletonCallSite singletonCallSite)
+            {
+                // If root call site is singleton we can return Func calling
+                // _runtimeResolver.Resolve directly and avoid Expression generation
+                if (TryResolveSingletonValue(singletonCallSite, out var value))
+                {
+                    return scope => value;
+                }
+
+                return scope => _runtimeResolver.Resolve(callSite, scope);
+            }
+
+            return BuildType(callSite);
+        }
+
+        protected override Expression VisitTransient(TransientCallSite transientCallSite, ILEmitResolverBuilderContext argument)
+        {
+            var shouldCapture = BeginCaptureDisposable(transientCallSite.ServiceCallSite.ImplementationType, argument);
+
+            VisitCallSite(transientCallSite.ServiceCallSite, argument);
+
+            if (shouldCapture)
+            {
+                EndCaptureDisposable(argument);
+            }
+            return null;
+        }
+
+        protected override Expression VisitConstructor(ConstructorCallSite constructorCallSite, ILEmitResolverBuilderContext argument)
+        {
+            foreach (var parameterCallSite in constructorCallSite.ParameterCallSites)
+            {
+                VisitCallSite(parameterCallSite, argument);
+            }
+            argument.Generator.Emit(OpCodes.Newobj, constructorCallSite.ConstructorInfo);
```

Comment:
Do we support value types?

Reply:
Has anyone ever put a struct in the container? Now I'm curious...

Reply:
I'm sure someone did and we have to support it.

Reply:
Make sure we add tests for it.

Reply:
It didn't work in expression either, I tried adding tests and they fail.

Reply:
Yeah but the IL for newing up a struct is different soooo

Reply:
Nono, I'm sure they would fail in my code. They are failing in old Expression generating code too.